src/lib/GenericPredictor.cc

```diff
@@ -70,13 +72,23 @@ BranchPrediction GenericPredictor::predict(uint64_t address, BranchType type,
   } else if (type == BranchType::Conditional) {
     if (!prediction.taken) prediction.target = address + 4;
   }
+
+  // Store the hashed index for correct hashing in update()
+  ftq_.emplace_back(prediction.taken, hashedIndex);
+
+  // Speculatively update the global history
+  globalHistory_ =
+      ((globalHistory_ << 1) | prediction.taken) & globalHistoryLength_;
+
   return prediction;
 }
 
 void GenericPredictor::update(uint64_t address, bool taken,
                               uint64_t targetAddress, BranchType type) {
-  // Get previous index calculated for the instruction address supplied
-  uint64_t hashedIndex = btbHistory_[address];
+  // Get previous prediction and index calculated from the FTQ
+  bool prevPrediction = ftq_.front().first;
+  uint64_t hashedIndex = ftq_.front().second;
+  ftq_.pop_front();
```

Comment:
The FTQ stores the relevant state information of the BP for updating the BP from the time of prediction.  This is needed to correctly update the BP.  It is an improvement on the previous solution in that it can handle multiple instances of the same branch instruction being in flight at the same time.  Further, the FTQ allows speculative execution by keeping a track of the number of branches in flight (as you note).  

As shown by the branch prediction metrics in the description of this PR, the FTQ improves the branch predictor's prediction rate.  Therefore, it must be doing something.  

I agree that it adds complexity (though would query "a lot of"), but I think complexity is needed if we want to have an effective branch predictor.